experimental/iterators/lib/Dialect/Iterators/Transforms/DecomposeIteratorStates.cpp

```diff
@@ -0,0 +1,251 @@
+
+#include "iterators/Dialect/Iterators/IR/Iterators.h"
+#include "iterators/Dialect/Iterators/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/OneToNFuncConversions.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/OneToNTypeConversion.h"
+
+namespace mlir {
+#define GEN_PASS_CLASSES
+#include "iterators/Dialect/Iterators/Transforms/Passes.h.inc"
+} // namespace mlir
+
+// using namespace iterators;
+using namespace mlir;
+using namespace mlir::iterators;
+
+class DecomposeStatesTypeConverter : public OneToNTypeConverter {
+public:
+  DecomposeStatesTypeConverter() {
+    addConversion([](Type type) { return type; });
+    addConversion([&](StateType type, SmallVectorImpl<Type> &results) {
+      return decomposeStateType(type, results);
+    });
+  }
+
+private:
+  /// Maps a StateType to the types of its fields.
+  Optional<LogicalResult> decomposeStateType(StateType type,
+                                             SmallVectorImpl<Type> &results) {
+    for (Type fieldType : type.getFieldTypes()) {
+      if (failed(convertTypes(fieldType, results)))
+        return failure();
+    }
+    return success();
+  }
+};
+
+class DecomposeCreateStateOp : public OneToNOpConversionPattern<CreateStateOp> {
+public:
+  using OneToNOpConversionPattern<CreateStateOp>::OneToNOpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(CreateStateOp op, OneToNPatternRewriter &rewriter,
+                  const OneToNTypeMapping & /*operandMapping*/,
+                  const OneToNTypeMapping &resultMapping,
+                  const ValueRange convertedOperands) const override {
+    // Simply replace the current op with the converted operands.
+    rewriter.replaceOp(op, convertedOperands, resultMapping);
+    return success();
+  }
+};
+
+class DecomposeInsertValueOp : public OneToNOpConversionPattern<InsertValueOp> {
+public:
+  using OneToNOpConversionPattern<InsertValueOp>::OneToNOpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(InsertValueOp op, OneToNPatternRewriter &rewriter,
+                  const OneToNTypeMapping &operandMapping,
+                  const OneToNTypeMapping &resultMapping,
+                  const ValueRange convertedOperands) const override {
+    // Construct conversion mapping for field types.
+    auto stateType = op.getState().getType().cast<StateType>();
+    TypeRange originalFieldTypes = stateType.getFieldTypes();
+    OneToNTypeMapping fieldMapping(originalFieldTypes);
+    if (failed(typeConverter->convertSignatureArgs(originalFieldTypes,
+                                                   fieldMapping)))
+      return failure();
+
+    // Extract converted operands.
+    ValueRange convertedState =
+        operandMapping.getConvertedValues(convertedOperands, 0);
```

Comment:
is there an opportunity to use the Adaptor to get to the magic constant index?

Reply:
Yeah, that'd be nice. This is an aspect where the 1:N type conversion could be improved. However, I think that the operand indices are currently not exposed. The are *used* by the accessor methods but they aren't exposed.

What would also be nice is if I could create an adaptor with the converted operands insight like the traditional dialect conversion does -- but then an accessor (like `getState()`) would have to return a value *range* rather than a single value. Is that something you think I should investigate/push for?

Reply:
Ha! I looked at the adaptors again: they look more powerful from what I remember from when I looked last time. There is (now?) a templated `XOpGenericAdaptor`, which you can wrap around a range of anything and then have the accessors access that. I believe can wrap that around a range of ranges such that the accessor returns the *range* of converted values. That would be an awesome solution upstream. I'll report back this afternoon.

Reply:
OK, this issue will be fixed upstream if https://reviews.llvm.org/D147225 is accepted. Can we allow this PR with the magic constants and improve upon it when we integrate the upstream changes?